addOns/exim/src/main/java/org/zaproxy/addon/exim/har/HarImporter.java

```diff
@@ -118,29 +143,128 @@ private static void setHttpResponse(HarResponse harResponse, HttpMessage message
         }
     }
 
-    private void importHarFile(File file) {
+    private void importHarFile(HarLog log) {
         try {
-            processMessages(file);
+            processMessages(log);
             Stats.incCounter(ExtensionExim.STATS_PREFIX + STATS_HAR_FILE);
             success = true;
         } catch (IOException e) {
             LOGGER.warn(
-                    Constant.messages.getString(
-                            ExtensionExim.EXIM_OUTPUT_ERROR, file.getAbsolutePath()));
+                    Constant.messages.getString(ExtensionExim.EXIM_OUTPUT_ERROR, e.getMessage()));
             LOGGER.warn(e);
             Stats.incCounter(ExtensionExim.STATS_PREFIX + STATS_HAR_FILE_ERROR);
             success = false;
         }
     }
 
-    private void processMessages(File file) throws IOException {
-        List<HttpMessage> messages =
-                HarImporter.getHttpMessages(new HarFileReader().readHarFile(file));
-        int count = 1;
+    private HarLog preProcessHarLog(HarLog log) {
+        List<HarEntry> forRemoval = new ArrayList<>();
+        List<HarEntry> localPrivate = new ArrayList<>();
+        log.getEntries().stream()
+                .forEach(entry -> preProcessHarEntry(forRemoval, localPrivate, entry));
+        log.getEntries().removeAll(forRemoval);
+        log.getEntries().removeAll(localPrivate);
+        int adjustment = forRemoval.size() + localPrivate.size();
+        updateProgress(adjustment, "");
+        return log;
+    }
+
+    private void preProcessHarEntry(
+            List<HarEntry> forRemoval, List<HarEntry> localPrivate, HarEntry entry) {
+        preProcessHttpVersion(forRemoval, entry);
+
+        String url = entry.getRequest().getUrl();
+        preProcessLocalPrivateEntries(localPrivate, entry, url);
+
+        // No point processing headers on entries being dropped
```

Comment:
Would be better to filter the stream and this would happen naturally.

Reply:
You mean within the main loop discard/drop the messages one by one instead of pre-processing?

Reply:
If you use the filter method of the stream only the messages that pass the filter are processed further.

Reply:
It's already using a stream just not using the filter.

Reply:
We're probably talking about different locations. But all good. I'll go through it and work it out.